src/forwarder/client.rs

```diff
@@ -0,0 +1,115 @@
+        // deployments
+        // .http2_prior_knowledge()
+        // ---------------------------
+        //The initial size of a single stream, which corresponds to a single request. It defaults
+        //to 64KiB. Bumping it increases memory usage (not a problem), but most importantly it
+        //increases the TCP send socket buffer usage of a given stream (remember there is a single
+        //connection behind it). However, there are many “big requests” to forward in the proxy
+        //above 50KiB, so to be safe we could set it to a value like 128KiB or 256KiB.
+        // .http2_initial_stream_window_size(HTTP2_INITIAL_STREAM_WINDOW_SIZE)
+        // Defines the total number of bytes that can be outstanding (sent but unacknowledged)
+        // across all streams in a single HTTP/2 connection. By default (as of RFC 7540),
+        // default size is 64KiB. This is too low for any use-case like ours, so we should
+        // definitely bump it probably to be a bit higher (1.5x/2x) than BDP for that peer.
+        // Given that we don’t discriminate by expected RTT, we have to find a common value
+        // for all of them as of now.
+        // .http2_initial_connection_window_size(HTTP2_INITIAL_CONNECTION_WINDOW_SIZE as u32)
+        // Adaptively adjust stream window size based on responses received, to be closer to BDP. I
+        // don’t know if we need it, we should look at the reqwest implementation what it
+        // does exactly to avoid weird surprises.
+        .http2_adaptive_window(true)
+        // Sets the maximum number of bytes allowed in the payload of a frame. Defaults to 16,384
+        // bytes, HAProxy doesn’t recommend to change it and probably we should not do it even here.
+        .http2_max_frame_size(16_384)
+        // Related to header size, we can leave the default of 16KiB.
+        .http2_max_header_list_size(16_384)
+        // Interval to send ping frames to keep the HTTP connection alive. By default it is
+        // disabled, however we should a message to a peer multiple times per second so the
+        // connection should never become idle. We can still set it to something like 10s if
+        // needed.
+        .http2_keep_alive_interval(Duration::from_secs(10))
+        // By default disabled, we don’t want to close connections so we can avoid it.
+        // .http2_keep_alive_timeout(timeout)
+        // ----------------------------
+        // Sets whether HTTP2 keep-alive should apply while the connection is idle. Again we could
+        // enable it, but the connection should never become idle with HTTP/2.
+        .http2_keep_alive_while_idle(true);
+    builder
+}
+
+#[derive(Debug, Clone)]
+pub struct ClientPool {
+    /// The clients in the pool.
+    clients: Arc<[reqwest::Client]>,
+    /// The number of clients in the pool.
+    num_clients: usize,
+    /// The index of the last used client. Used for round-robin load balancing.
+    last_used: Arc<AtomicU8>,
```

Comment:
On a better thought, we could load balance based on the number of concurrent streams open within a certain connection. But as a first iteration this works perfectly fine.